File: src/1463_CherryPickupII.cpp
#include "utils.hpp"


class Solution {
  public:
    int cherryPickup(vector<vector<int>>& grid) {
      int m = grid.size(), n = grid[0].size();

      vector<vector<vector<int>>> memo(m+1, vector<vector<int>>(n, vector<int>(n, 0)));
      // int memo[m+1][n][n];
      // memset(memo, 0, (m+1)*n*n);
      
      int dir[3] = {-1, 0, 1};
      for (int k=m; k>0; --k) {
        for (int i=0; i<n-1; ++i)
          for (int j=i+1; j<n; ++j) 
            for (int l=0; l<9; ++l) {
              int i_prev = i+dir[l/3];
              int j_prev = j+dir[l%3];

              if (i_prev>=j_prev || i_prev<0 || j_prev>=n) 
                continue;

              memo[k-1][i_prev][j_prev] = max(memo[k-1][i_prev][j_prev], memo[k][i][j]+grid[k-1][i_prev]+grid[k-1][j_prev]);
            }
      }

      return memo[0][0][n-1];
    }
};


int main() {
  UFUNC(Solution::cherryPickup);
  return 0;
}

File: src/1463_CherryPickupII.java
class CherryPickupII {
  int m, n;
  int[] dir;
  int[][] grid;
  int[][][] memo;
  boolean[][][] seen;
  public int cherryPickup(int[][] grid) {
    m = grid.length;
    n = grid[0].length;
    this.grid = grid;

    dir = new int[] {-1, 0, 1};
    memo = new int[m][n][n];
    seen = new boolean[m][n][n];
    return dp(0, 0, n-1);
  }

  int dp(int i, int j, int k) {
    if (i==m || j<0 || k<0 || j>=n || k>=n) return 0;
    if (seen[i][j][k]) return memo[i][j][k];
    int res = 0;

    for (int p=0; p<3; ++p)
      for (int q=0; q<3; ++q) {
        int next_j = j + dir[p], next_k = k + dir[q];
        if (next_j>next_k) continue;
        res = Math.max(res, dp(i+1, next_j, next_k));
      }

    res += grid[i][j];
    if (j!=k)
      res += grid[i][k];

    seen[i][j][k] = true;
    memo[i][j][k] = res;
    return res;
  }
}